aten/src/ATen/CPUApplyUtils.h

```diff
@@ -0,0 +1,271 @@
+  int64_t TH_TENSOR_dim_index = 0; \
+  __ATH_TENSOR_APPLYX_PREAMBLE(TYPE, ATENSOR1, DIM, 1) \
+  __ATH_TENSOR_APPLYX_PREAMBLE(TYPE, ATENSOR2, DIM, 1) \
+\
+  auto t1_numel = ATENSOR1.numel(); \
+  auto t2_numel = ATENSOR2.numel(); \
+  if(t1_numel != t2_numel) {                                    \
+    std::ostringstream oss; \
+    oss << "inconsistent tensor size, expected " << ATENSOR1.sizes() << " and " << ATENSOR2.sizes() \
+        << " to have the same number of elements, but got " << t1_numel << " and " << t2_numel << " elements respectively"; \
+    throw std::runtime_error(oss.str()); \
+  }                                                                   \
+  while(!TH_TENSOR_APPLY_hasFinished) \
+  { \
+    /* Loop through the inner most region of the Tensor */ \
+    for(; ATENSOR1##_i < ATENSOR1##_size && ATENSOR2##_i < ATENSOR2##_size; ATENSOR1##_i++, ATENSOR2##_i++, ATENSOR1##_data += ATENSOR1##_stride, ATENSOR2##_data += ATENSOR2##_stride) /* 0 et pas TENSOR##_dim! */ \
+    { \
+      CODE \
+    } \
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(ATENSOR1, 0) \
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(ATENSOR2, 0) \
+  } \
+  if(ATENSOR1##_counter != NULL) \
+    delete [] ATENSOR1##_counter; \
+  if(ATENSOR2##_counter != NULL) \
+    delete [] ATENSOR2##_counter; \
+}
+
+template <typename ScalarType, typename Op>
+void CPU_tensor_apply2_dim(Tensor& tensor1, Tensor& tensor2, int64_t dim, Op op) {
+  check_correct_backend(tensor1, tensor2);
+  bool TH_TENSOR_APPLY_hasFinished = false;
+  int64_t TH_TENSOR_dim_index = 0;
+  __ATH_TENSOR_APPLYX_PREAMBLE(ScalarType, tensor1, dim, 1)
+  __ATH_TENSOR_APPLYX_PREAMBLE(ScalarType, tensor2, dim, 1)
+  auto t1_numel = tensor1.numel();
+  auto t2_numel = tensor2.numel();
+  if(t1_numel != t2_numel) {
+    std::ostringstream oss;
+    oss << "inconsistent tensor size, expected " << tensor1.sizes() << " and " << tensor2.sizes()
+        << " to have the same number of elements, but got " << t1_numel << " and " << t2_numel << " elements respectively";
+    throw std::runtime_error(oss.str());
+  }
+  while(!TH_TENSOR_APPLY_hasFinished)
+  {
+    /* Loop through the inner most region of the Tensor */
+    for(; tensor1_i < tensor1_size && tensor2_i < tensor2_size; tensor1_i++, tensor2_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride)
+    {
+      op(*tensor1_data, *tensor2_data, TH_TENSOR_APPLY_hasFinished);
+    }
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor1, 0)
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor2, 0)
+  }
+  if(tensor1_counter != NULL)
+    delete [] tensor1_counter;
+  if(tensor2_counter != NULL)
+    delete [] tensor2_counter;
+}
+
+template<typename ScalarType, typename Op>
+void CPU_tensor_apply2(Tensor tensor1, Tensor tensor2, Op op) {
+  CPU_tensor_apply2_dim<ScalarType, Op>(tensor1, tensor2, -1, op);
+}
+
+template <typename ScalarType, typename Op>
+void CPU_tensor_apply3_dim(Tensor &tensor1, Tensor& tensor2, Tensor& tensor3, int64_t dim, Op op) {
```

Comment:
Having read this code a bit more carefully, it does seem plausible that you could templatize it without too much more duplication. For essentially of the preamble (excluding the loop), it seems you could simply create a struct which initializes all of the per tensor variables for one tensor, and iterate over that (using a compile-time integer) per tensor. Then the final problem is the actual iteration but a simple recursive variadic template will work; something like here https://godbolt.org/g/vqRBTG

Of course, the danger of all of this is that the compiler won't be able to figure it out and things get slower, but it seems to do OK on the simple example!

Reply:
seems plausible; I want to have some reasonable benchmarking in place before I embark on that; I was doing things on the python level but this only supports Variables, the older version only Tensors (unless you use the python function and swallow that overhead).  So I'll just do the benchmarking at the C++ level.